web/packages/teleterm/src/ui/HeadlessAuthn/HeadlessPrompt/HeadlessPrompt.tsx

```diff
@@ -42,14 +44,22 @@ export type HeadlessPromptProps = {
 export function HeadlessPrompt({
   cluster,
   clientIp,
+  skipConfirm,
   onApprove,
+  abortApproval,
   onReject,
   headlessAuthenticationId,
   updateHeadlessStateAttempt,
   onCancel,
 }: HeadlessPromptProps) {
   const [waitForMfa, setWaitForMfa] = useState(false);
 
+  // skip to MFA confirmation step.
+  if (skipConfirm) {
+    setWaitForMfa(true);
+    onApprove();
```

Comment:
@ravicious @gzdunek For some reason, this `OnApprove` call results in an infinite loop, sending headless approval requests to the tshd daemon and overloading the yubikey. I have no idea why this would be an infinite loop while the on button press below works perfectly fine. Any ideas?

Reply:
That's because the body of a component function, `HeadlessPrompt`, is executed on each render and a component gets rendered each time you update it's state. So you end up with an infinite loop.

Try this patch:

<details>
<summary>Patch</summary>

```diff
diff --git a/web/packages/teleterm/src/ui/HeadlessAuthn/HeadlessAuthentication.tsx b/web/packages/teleterm/src/ui/HeadlessAuthn/HeadlessAuthentication.tsx
index 08da0c1aa6..fc73d9b1c3 100644
--- a/web/packages/teleterm/src/ui/HeadlessAuthn/HeadlessAuthentication.tsx
+++ b/web/packages/teleterm/src/ui/HeadlessAuthn/HeadlessAuthentication.tsx
@@ -14,7 +14,7 @@
  * limitations under the License.
  */
 
-import React, { useRef } from 'react';
+import React, { useRef, useEffect } from 'react';
 
 import { useAsync } from 'shared/hooks/useAsync';
 
@@ -69,6 +69,12 @@ export function HeadlessAuthentication(props: HeadlessAuthenticationProps) {
     }
   }
 
+  useEffect(() => {
+    if (props.skipConfirm && updateHeadlessStateAttempt.status === '') {
+      handleHeadlessApprove();
+    }
+  }, []);
+
   return (
     <HeadlessPrompt
       cluster={cluster}
diff --git a/web/packages/teleterm/src/ui/HeadlessAuthn/HeadlessPrompt/HeadlessPrompt.tsx b/web/packages/teleterm/src/ui/HeadlessAuthn/HeadlessPrompt/HeadlessPrompt.tsx
index f1da2f1bd8..b0294cb162 100644
--- a/web/packages/teleterm/src/ui/HeadlessAuthn/HeadlessPrompt/HeadlessPrompt.tsx
+++ b/web/packages/teleterm/src/ui/HeadlessAuthn/HeadlessPrompt/HeadlessPrompt.tsx
@@ -52,13 +52,9 @@ export function HeadlessPrompt({
   updateHeadlessStateAttempt,
   onCancel,
 }: HeadlessPromptProps) {
-  const [waitForMfa, setWaitForMfa] = useState(false);
-
-  // skip to MFA confirmation step.
-  if (skipConfirm) {
-    setWaitForMfa(true);
-    onApprove();
-  }
+  // skipConfirm automatically attempts to approve a headless auth attempt,
+  // so let's show waitForMfa from the very beginning in that case.
+  const [waitForMfa, setWaitForMfa] = useState(skipConfirm);
 
   return (
     <DialogConfirmation
```
</details>

Reply:
Ah I suspected it was something like that but had no idea how to fix it, thanks!